kura/org.eclipse.kura.wire.camel/src/main/java/org/eclipse/kura/wire/camel/CamelProduce.java

```diff
@@ -9,24 +9,97 @@
  * 
  * Contributors:
  *  Red Hat Inc
+ *  heyoulin <heyoulin@gmail.com>
  *******************************************************************************/
 package org.eclipse.kura.wire.camel;
 
 import static org.apache.camel.builder.DefaultFluentProducerTemplate.on;
 
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
 import org.apache.camel.CamelContext;
+import org.apache.camel.FluentProducerTemplate;
 import org.eclipse.kura.wire.WireEnvelope;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class CamelProduce extends AbstractReceiverWireComponent {
 
+    private static final Logger logger = LoggerFactory.getLogger(CamelProduce.class);
+
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock(true);
+
+    private FluentProducerTemplate template = null;
+
     @Override
     protected void processReceive(final CamelContext context, final String endpointUri, final WireEnvelope envelope)
             throws Exception {
 
-        on(context) //
-                .withBody(envelope) //
-                .to(endpointUri) //
-                .asyncSend();
+        boolean templateAvailable = false;
+        final Lock rlock = this.rwLock.readLock();
+        rlock.lock();
+        try {
+            templateAvailable = isTemplateAvailable(context);
+        } finally {
+            rlock.unlock();
+        }
+        if (!templateAvailable) {
+            createTemplate(context);
+        }
+        if (template != null) {
```

Comment:
I know that. Because i don't want sending messages within a read lock. 

If no sychronization we can discard this messages. It is acceptable.

If switch template the messages is unuseful. We can discard it and wait next message.

This PR is mainly to solve the thread leak problem instead of modifying the settings causing message loss

Reply:
In AbstractCamelWireComponent and other componet there have many same problem.

For runtime efficiency these errors are acceptable and can automatically recoverable at next message comming. And these error are also justified. We can just debug or info.
But a memory leak is unacceptable.

Reply:
Lots of locks maybe cause system dead lock or slow down the overall system efficiency

Reply:
In my opinion the code should still be the same as the original. just add try catch to skip processReceive massage

Only createTemplate and deactivate add synchronized keyword. These two write methods are rarely used.

How the code is refactored is up to you.

'''

public class CamelProduce extends AbstractReceiverWireComponent {

    private static final Logger logger = LoggerFactory.getLogger(CamelProduce.class);

    private FluentProducerTemplate template = null;

    @Override
    protected void processReceive(final CamelContext context, final String endpointUri, final WireEnvelope envelope)
            throws Exception {

        if (!isTemplateAvailable(context)) {
            createTemplate(context);
        }
        if (template != null) {
            try {
                template //
                        .withBody(envelope) //
                        .to(endpointUri) //
                        .asyncSend();
            } catch (Exception e) {
                logger.error("asyncSend error", e);
            }
        } else {
            logger.debug("FluentProducerTemplate is changing. Skip send massage and wait next massage");
        }
    }

    @Override
    protected synchronized void deactivate() {
        super.deactivate();
        closeTemplate();
    }

    private boolean isTemplateAvailable(final CamelContext context) {
        return template != null && template.getCamelContext() == context;
    }

    private synchronized void createTemplate(final CamelContext context) {
        closeTemplate();
        template = on(context);
    }

    private void closeTemplate() {
        if (template != null) {
            try {
                template.close();
            } catch (Exception ignored) {
            } finally {
                template = null;
            }
        }
    }

}

'''